### DISCUSSION TITLE
Data structure for road networks

### DISCUSSION THREAD

**Discussion by hkayabilisim** [TomorrowsCities/tomorrowscities#2] [General]
As of [v0.1](https://github.com/TomorrowsCities/tomorrowcities/releases/tag/v0.1), the data structure of the transportation network is given below. Let's hear some feedback and discuss possible ways to improve upon.

**Nodes**

| |node_id	| geometry|
|-|---------|---------|
|0|3385919019|	POINT (82.61515 27.80731)|
|1|3385919022|	POINT (82.61096 27.80733)|
|2|3385919040|	POINT (82.60816 27.80784)|
|3|3385919056|	POINT (82.60709 27.80830)|
|4|3385919116|	POINT (82.69962 27.80961)|

**Edges**
| |edge_id|	from_node|	to_node|	length|	bridge|	bridge_type|	geometry|
|-|-------|----------|---------|--------|-------|------------|----------|
|0|	0	|3385919019|	3762934344|	206.598	|False|	NaN|	LINESTRING (82.61515 27.80731, 82.61477 27.807...|
|1|	1	|3385919019|	3762934320|	954.306	|False|	NaN|	LINESTRING (82.61515 27.80731, 82.61585 27.807...|
|2|	2	|3385919019|	3762934257|	419.059	|False|	NaN|	LINESTRING (82.61515 27.80731, 82.61514 27.807...|
|3|	3	|3385919022|	3762934349|	26.608	|True	|Steel|	LINESTRING (82.61096 27.80733, 82.61082 27.807...|
|4|	4	|3385919022|	3762934344|	207.798	|False|	NaN	|LINESTRING (82.61096 27.80733, 82.61114 27.807...|

**Comment by hkayabilisim** [comment]
Here is a simple script to fetching existing road map from OpenStreetMap (OSM). It simply fetches the data from OSM, creates the graph, discard redundant attributes and finally create GeoJSON formatted node and edges compatible with the engine.  It might be useful for you @PrashantR-79 for Rapti application.

~~~python
#%%
import os
os.environ['USE_PYGEOS'] = '0'
import geopandas as gpd
import osmnx
import matplotlib.pyplot as plt
import contextily as ctx


place_name = "Sano Sisahaniya, Rapti, Dang, Lumbini Province, Nepal"
graph = osmnx.graph_from_address(place_name, dist=5000)
nodes, edges = osmnx.graph_to_gdfs(graph)

nodes = nodes.reset_index()
edges = edges.reset_index()

edges['edge_id']=edges.index
edges = edges.rename(columns={"u": "from_node", 
                        "v": "to_node"})
edges['bridge'] = edges['bridge'].apply(lambda x: True if  x=='yes' else False)
edges['bridge_type'] = edges['bridge'].apply(lambda x: 'MSC_concrete' if  x else None)
edges = edges[['edge_id','from_node','to_node','length','bridge','bridge_type','geometry']]
nodes = nodes.rename(columns={"osmid": "node_id"})
nodes = nodes[['node_id','geometry']]

fig, ax = plt.subplots(nrows=1,ncols=1,figsize=(10,10))
fig.set_tight_layout(True)
nodes.plot(ax=ax)
edges.plot(ax=ax)
edges[edges['bridge']].plot(ax=ax,color='red')
ax.set_title('Road Network')
ax.set_yticks([])
# Fetch a map layer from openstreetmap (not working in Kaggle)
ctx.add_basemap(ax,crs=nodes.crs,source=ctx.providers.OpenStreetMap.Mapnik)


nodes.to_file('rapti_road_nodes.geojson', driver='GeoJSON')
edges.to_file('rapti_road_edges.geojson', driver='GeoJSON')
~~~

**Reply by hkayabilisim** [reply]
Instead of **osmnx.graph_from_address(place_name)**, you can also use **osmnx.graph_from_polygon(polygon)** where polygon is the polygon containing the area of interest.

**Reply by PrashantR-79** [reply]
I used the above code with a few modifications for polygon based download. The code is attached along with sample polygon.
[osmnx_code.zip](https://github.com/TomorrowsCities/tomorrowscities/files/14022678/osmnx_code.zip)
I replaced "lambda x: 'MSC_concrete' if  x else None" with "lambda x: 'unknown' if  x else None"

**Reply by PrashantR-79** [reply]
Sample Output for the above example:
[Sample_Output.zip](https://github.com/TomorrowsCities/tomorrowscities/files/14022168/Sample_Output.zip)
